presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -334,14 +336,86 @@ private void enforceCpuLimits()
     {
         for (QueryExecution query : queryTracker.getAllQueries()) {
             Duration cpuTime = query.getTotalCpuTime();
+            Optional<ResourceGroupQueryLimits> resourceGroupLimit = query.getQueryLimits();
             Duration sessionLimit = getQueryMaxCpuTime(query.getSession());
-            Duration limit = Ordering.natural().min(maxQueryCpuTime, sessionLimit);
+            LimitWithLevelContext limitContext = compareDurationLimits(sessionLimit, extractCpuLimit(resourceGroupLimit), maxQueryCpuTime);
+            Duration limit = limitContext.getLimit();
             if (cpuTime.compareTo(limit) > 0) {
-                query.fail(new ExceededCpuLimitException(limit));
+                String limitSource;
+                switch (limitContext.getLimitLevel()){
+                    case QueryLevel:
+                        limitSource = "query level";
+                        break;
+                    case SystemLevel:
+                        limitSource = "system level";
+                        break;
+                    case ResourceGroupLevel:
+                        limitSource = "resource group level";
+                        break;
```

Comment:
nit - instead of having this separate switch case - every Enum can also store member variables. so you can store this string inside the Enum